src/node/state.rs

```diff
@@ -0,0 +1,264 @@
+            PbftPhase::Preparing => "Pr",
+            PbftPhase::Checking => "Ch",
+            PbftPhase::Committing => "Co",
+            PbftPhase::FinalCommitting => "FC",
+            PbftPhase::Finished => "Fi",
+        };
+
+        let wb = match self.working_block {
+            WorkingBlockOption::WorkingBlock(ref block) =>
+                String::from(&hex::encode(block.get_block_id())[..6]),
+            WorkingBlockOption::WorkingBlockNew(ref block_id) =>
+                String::from(&hex::encode(block_id)[..5]) + "~",
+            _ =>
+                String::from("~none~"),
+        };
+
+        write!(
+            f,
+            "({} {} {}, seq {}, wb {}), Node {}{:02}",
+            phase, mode, self.view, self.seq_num, wb, ast, self.id,
+        )
+    }
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub enum WorkingBlockOption {
+    NoWorkingBlock,
+    WorkingBlockNew(BlockId),
+    WorkingBlock(PbftBlock),
+}
+
+impl WorkingBlockOption {
+    pub fn is_none(&self) -> bool {
+        self == &WorkingBlockOption::NoWorkingBlock
+    }
+
+    pub fn is_some(&self) -> bool {
+        match self {
+            &WorkingBlockOption::WorkingBlock(_) => true,
+            _ => false,
+        }
+    }
+}
+
+// Information about the PBFT algorithm's state
+#[derive(Debug)]
+pub struct PbftState {
+    // This node's ID
+    pub id: u64,
+
+    // The node's current sequence number
+    // Always starts at 0; representative of an unknown sequence number.
+    pub seq_num: u64,
+
+    // The current view (where the primary's ID is p = v mod network_node_ids.len())
+    pub view: u64,
+
+    // Current phase of the algorithm
+    pub phase: PbftPhase,
+
+    // Is this node primary or secondary?
+    role: PbftNodeRole,
+
+    // Normal operation, view change, or checkpointing. Previous mode is stored when checkpointing
+    pub mode: PbftMode,
+    pub pre_checkpoint_mode: PbftMode,
+
+    // Map of peers in the network, including ourselves
+    network_node_ids: HashMap<u64, PeerId>,
+
+    // The maximum number of faulty nodes in the network
+    pub f: u64,
+
+    // Timer used to keep track of whether or not this node has received timely messages from the
+    // primary. If a message hasn't been received in a certain amount of time, then this node will
+    // initiate a view change.
+    pub timeout: Timeout,
+
+    // The current block we're working on
+    pub working_block: WorkingBlockOption,
+}
+
+impl PbftState {
+    pub fn new(id: u64, config: &PbftConfig) -> Self {
+        let peer_id_map: HashMap<u64, PeerId> = config
+            .peers
+            .clone()
+            .into_iter()
+            .map(|(peer_id, node_id)| (node_id, peer_id))
+            .collect();
+
+        // Maximum number of faulty nodes in this network
+        let f = ((peer_id_map.len() - 1) / 3) as u64;
+        if f == 0 {
+            warn!("This network does not contain enough nodes to be fault tolerant");
```

Comment:
What happens if there are not enough nodes?

Reply:
In other implementations I've seen, nodes will exit status 1 when the network doesn't have enough

Reply:
Could the network pause (retry the population check) until there is a large enough population?

Reply:
Currently, this implementation only supports a static network, so I think it might make more sense to either panic or exit status 1 here (for now). Once more intelligent networking is developed, I agree that pausing until the network is large enough would be good.